File: typescript/src/bc/big-tofu.ts
/** OpenTofu/Terraform construct helpers. Ported from big-tofu.core. */

/** Sanitize a `ns/name` keyword string into a Terraform-safe identifier. */
export function fqnToName(fqn: string, sep = "_"): string {
  const sanitize = (s: string) => s.replace(/[-.]/g, sep);
  const slash = fqn.indexOf("/");
  const ns = slash >= 0 ? sanitize(fqn.slice(0, slash)) : null;
  const name = sanitize(slash >= 0 ? fqn.slice(slash + 1) : fqn);
  return (ns ?? "") + (ns ? sep : "") + name;
}

/** Append a suffix to the `name` part of a `ns/name` keyword string. */
export function addSuffix(fqn: string, suffix: string): string {
  return fqn + suffix;
}

export interface Construct {
  group: string;
  type: string;
  fqn: string;
  block: Record<string, any>;
}

/** Build a Construct (Terraform resource/data descriptor). */
export function makeConstruct(
  group: string,
  type: string,
  fqn: string,
  block: Record<string, any>,
): Construct {
  return { group, type, fqn, block };
}

/** Render a Construct into its nested `{group: {type: {name: block}}}` map. */
export function construct(c: Construct): Record<string, any> {
  return { [c.group]: { [c.type]: { [fqnToName(c.fqn)]: c.block } } };
}

File: src/once/bc/big_tofu.py
"""OpenTofu/Terraform construct helpers."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


def fqn_to_name(fqn: str, sep: str = "_") -> str:
    """Sanitize a ``ns/name`` keyword into a Terraform-safe identifier."""

    def sanitize(s: str) -> str:
        return s.replace("-", sep).replace(".", sep)

    if "/" in fqn:
        ns, name = fqn.split("/", 1)
        return f"{sanitize(ns)}{sep}{sanitize(name)}"
    return sanitize(fqn)


def add_suffix(fqn: str, suffix: str) -> str:
    """Append a suffix to the name part (matching the Clojure helper behavior)."""
    return f"{fqn}{suffix}"


@dataclass(frozen=True)
class Construct:
    group: str
    type: str
    fqn: str
    block: dict[str, Any]


def make_construct(group: str, type_: str, fqn: str, block: dict[str, Any]) -> Construct:
    return Construct(group=group, type=type_, fqn=fqn, block=block)


def construct(c: Construct) -> dict[str, Any]:
    """Render a construct into ``{group: {type: {name: block}}}``."""
    return {c.group: {c.type: {fqn_to_name(c.fqn): c.block}}}


fqnToName = fqn_to_name
addSuffix = add_suffix
makeConstruct = make_construct
